runtime/vcache/loader.go

```diff
@@ -51,560 +34,24 @@ type loader struct {
-		l.loadVector(g, projection, s.values)
-	case *union:
-		l.loadUnion(g, projection, s)
-	case *int_:
-		l.loadInt(g, s)
-	case *uint_:
-		l.loadUint(g, s)
-	case *primitive:
-		l.loadPrimitive(g, s)
-	case *const_:
-		s.mu.Lock()
-		vec := s.vec
-		if vec == nil {
-			// Map the const super.Value in the csup's type context to
-			// a new one in the query type context.
-			val := s.meta.Value
-			typ, err := l.sctx.TranslateType(val.Type())
-			if err != nil {
-				panic(err)
-			}
-			vec = vector.NewConst(super.NewValue(typ, val.Bytes()), s.length(), s.nulls.flat)
-			s.vec = vec
-		}
-		s.mu.Unlock()
-	case *dict:
-		l.loadDict(g, projection, s)
-	case *error_:
-		l.loadVector(g, projection, s.values)
-	case *named:
-		l.loadVector(g, projection, s.values)
-	default:
-		panic(fmt.Sprintf("vector cache: shadow type %T not supported", s))
-	}
-}
-
-func (l *loader) loadRecord(g *errgroup.Group, projection field.Projection, s *record) {
-	if len(projection) == 0 {
-		// Load the whole record.  We're either loading all on demand (nil paths)
-		// or loading this record because it's referenced at the end of a projected path.
-		for _, f := range s.fields {
-			if f != nil {
-				l.loadVector(g, nil, f)
-			}
-		}
-		return
-	}
-	switch elem := projection[0].(type) {
-	case string:
-		if k := indexOfField(elem, s.meta); k >= 0 {
-			l.loadVector(g, projection[1:], s.fields[k])
-		}
-	case field.Fork:
-		// Multiple fields at this level are being projected.
-		for _, path := range elem {
-			// records require a field name path element (i.e., string)
-			if name, ok := path[0].(string); ok {
-				if k := indexOfField(name, s.meta); k >= 0 {
-					l.loadVector(g, projection[1:], s.fields[k])
-				}
-			}
-		}
-	default:
-		panic(fmt.Sprintf("bad path in vcache loadRecord: %T", elem))
-	}
-}
-
-func (l *loader) loadUnion(g *errgroup.Group, projection field.Projection, s *union) {
-	l.loadUint32(g, &s.mu, &s.tags, s.meta.Tags)
-	for _, val := range s.values {
-		l.loadVector(g, projection, val)
-	}
-}
-
-func (l *loader) loadInt(g *errgroup.Group, s *int_) {
-	s.mu.Lock()
-	if s.vec != nil {
-		s.mu.Unlock()
-		return
-	}
-	s.mu.Unlock()
-	g.Go(func() error {
-		s.mu.Lock()
-		defer s.mu.Unlock()
-		if s.vec != nil {
-			return nil
-		}
-		bytes := make([]byte, s.meta.Location.MemLength)
-		if err := s.meta.Location.Read(l.r, bytes); err != nil {
-			return err
-		}
-		vals := intcomp.UncompressInt64(byteconv.ReinterpretSlice[uint64](bytes), nil)
-		vals = extendForNulls(vals, s.nulls.flat, s.count)
-		typ, err := l.sctx.TranslateType(s.meta.Typ)
-		if err != nil {
-			panic(err)
-		}
-		s.vec = vector.NewInt(typ, vals, s.nulls.flat)
-		return nil
-	})
+	return s.project(l, projection), nil
 }
 
-func (l *loader) loadUint(g *errgroup.Group, s *uint_) {
-	s.mu.Lock()
-	if s.vec != nil {
-		s.mu.Unlock()
-		return
+func loadOffsets(r io.ReaderAt, slice *[]uint32, loc csup.Segment, length uint32, nulls bitvec.Bits) error {
```

Comment:
Nit: A return parameter would be more idiomatic than writing through a pointer parameter.
```suggestion
func loadOffsets(r io.ReaderAt, loc csup.Segment, length uint32, nulls bitvec.Bits) ([]uint32, error) {
```